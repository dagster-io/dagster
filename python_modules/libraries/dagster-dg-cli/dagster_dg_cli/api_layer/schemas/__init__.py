"""Pydantic schemas for Dagster Plus API."""
