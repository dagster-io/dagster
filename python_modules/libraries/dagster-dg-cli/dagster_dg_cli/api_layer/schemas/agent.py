"""Agent models for REST-like API."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class DgApiAgentStatus(str, Enum):
    """Agent status enum for FastAPI compatibility."""

    RUNNING = "RUNNING"
    STOPPED = "STOPPED"
    NOT_RUNNING = "NOT_RUNNING"
    UNHEALTHY = "UNHEALTHY"
    UNKNOWN = "UNKNOWN"


class DgApiAgentMetadataEntry(BaseModel):
    """Agent metadata key-value pair."""

    key: str
    value: str


class DgApiAgent(BaseModel):
    """Agent resource model."""

    id: str  # Agent IDs are strings in the GraphQL schema
    agent_label: Optional[str]  # Can be None in some cases
    status: DgApiAgentStatus
    last_heartbeat_time: Optional[float]
    metadata: list[DgApiAgentMetadataEntry]

    class Config:
        from_attributes = True  # For future ORM compatibility


class DgApiAgentList(BaseModel):
    """GET /api/agents response."""

    items: list[DgApiAgent]
    total: int
