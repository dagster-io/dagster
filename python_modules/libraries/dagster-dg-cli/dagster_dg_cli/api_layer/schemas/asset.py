"""Asset models for REST-like API."""

from typing import Optional

from pydantic import BaseModel


class DgApiAsset(BaseModel):
    """Asset resource model."""

    id: str
    asset_key: str  # "my/asset/key"
    asset_key_parts: list[str]  # ["my", "asset", "key"]
    description: Optional[str]
    group_name: str
    kinds: list[str]
    metadata_entries: list[dict]

    class Config:
        from_attributes = True


class DgApiAssetList(BaseModel):
    """GET /api/assets response."""

    items: list[DgApiAsset]
    cursor: Optional[str]  # Next cursor for pagination
    has_more: bool  # Whether more results exist
