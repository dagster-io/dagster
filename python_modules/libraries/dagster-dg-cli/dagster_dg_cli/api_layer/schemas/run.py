"""Run metadata schema definitions."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class DgApiRunStatus(str, Enum):
    """Run execution status."""

    QUEUED = "QUEUED"
    STARTING = "STARTING"
    STARTED = "STARTED"
    SUCCESS = "SUCCESS"
    FAILURE = "FAILURE"
    CANCELING = "CANCELING"
    CANCELED = "CANCELED"


class DgApiRun(BaseModel):
    """Single run metadata model."""

    id: str
    status: DgApiRunStatus
    created_at: float  # Unix timestamp (seconds since epoch)
    started_at: Optional[float] = None  # Unix timestamp (seconds since epoch)
    ended_at: Optional[float] = None  # Unix timestamp (seconds since epoch)
    job_name: Optional[str] = None

    class Config:
        from_attributes = True
