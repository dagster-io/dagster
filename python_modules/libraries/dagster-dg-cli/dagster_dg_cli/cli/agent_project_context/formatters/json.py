"""JSON formatter for project context."""

import json

from dagster_dg_cli.cli.agent_project_context.formatters import BaseFormatter
from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class JsonFormatter(BaseFormatter):
    """Formats project context as JSON."""

    def format(self, context: ProjectContext) -> str:
        """Format project context as JSON."""
        # Convert record objects to dictionaries for JSON serialization
        context_dict = {
            "working_directory": context.working_directory,
            "project_root": context.project_root,
            "project_name": context.project_name,
            "components": [
                {
                    "key": comp.key,
                    "summary": comp.summary,
                    "namespace": comp.namespace,
                    "name": comp.name,
                }
                for comp in context.components
            ],
            "integrations": [
                {
                    "name": integ.name,
                    "description": integ.description,
                    "pypi": integ.pypi,
                }
                for integ in context.integrations
            ],
            "project_structure": {
                "has_pyproject_toml": context.project_structure.has_pyproject_toml,
                "has_dbt_project": context.project_structure.has_dbt_project,
                "dbt_project_paths": context.project_structure.dbt_project_paths,
                "python_packages": context.project_structure.python_packages,
            },
            "dbt_schemas": {
                key: {
                    "description": schema.description,
                    "scaffold_params_schema": schema.scaffold_params_schema,
                    "component_schema": schema.component_schema,
                }
                for key, schema in context.dbt_schemas.items()
            },
        }
        return json.dumps(context_dict, indent=2)
