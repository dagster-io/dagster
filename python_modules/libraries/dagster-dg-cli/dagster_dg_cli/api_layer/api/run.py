"""Run metadata API implementation."""

from dataclasses import dataclass
from typing import TYPE_CHECKING

from dagster_dg_cli.api_layer.graphql_adapter.run import get_run_via_graphql
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.run import DgApiRun


@dataclass(frozen=True)
class DgApiRunApi:
    """API for run metadata operations."""

    client: IGraphQLClient

    def get_run(self, run_id: str) -> "DgApiRun":
        """Get run metadata by ID."""
        return get_run_via_graphql(self.client, run_id)
