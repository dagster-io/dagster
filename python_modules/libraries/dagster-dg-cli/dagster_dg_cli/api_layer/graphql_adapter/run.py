"""GraphQL adapter for run metadata."""

from dagster_dg_cli.api_layer.schemas.run import DgApiRun, DgApiRunStatus
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

RUN_METADATA_QUERY = """
query DgApiRunMetadataQuery($runId: ID!) {
    runOrError(runId: $runId) {
        __typename
        ... on Run {
            id
            runId
            status
            creationTime
            startTime
            endTime
            jobName
        }
        ... on RunNotFoundError {
            message
        }
        ... on PythonError {
            message
            stack
        }
    }
}
"""


def get_run_via_graphql(client: IGraphQLClient, run_id: str) -> DgApiRun:
    """Get run metadata via GraphQL."""
    variables = {"runId": run_id}

    # Try/catch here is a temporary solution until we fix the GQL-- client.execute is throwing an
    # error if the run is not found, instead of returning the error in the response.
    try:
        result = client.execute(RUN_METADATA_QUERY, variables)
    except Exception:
        raise Exception(f"Run not found: {run_id}")

    run_result = result.get("runOrError")
    if not run_result:
        raise Exception(f"Run not found: {run_id}")

    typename = run_result.get("__typename")

    # Handle GraphQL errors
    if run_result.get("__typename") != "Run":
        error_msg = run_result.get("message", f"Graphql error: {typename}")
        raise Exception(error_msg)

    return DgApiRun(
        id=run_result["runId"],
        status=DgApiRunStatus(run_result["status"]),
        created_at=run_result["creationTime"],
        started_at=run_result.get("startTime"),
        ended_at=run_result.get("endTime"),
        job_name=run_result.get("jobName"),
    )
