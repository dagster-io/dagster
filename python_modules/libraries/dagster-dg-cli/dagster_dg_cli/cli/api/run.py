"""Run API commands following GitHub CLI patterns."""

import json

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper
from dagster_shared.plus.config import DagsterPlusCliConfig
from dagster_shared.plus.config_utils import dg_api_options

from dagster_dg_cli.cli.api.client import create_dg_api_graphql_client


def format_run_table(run) -> str:
    """Format run as human-readable table."""
    lines = [
        f"Run ID: {run.id}",
        f"Status: {run.status}",
        f"Created: {run.created_at}",
    ]

    if run.started_at:
        lines.append(f"Started: {run.started_at}")
    if run.ended_at:
        lines.append(f"Ended: {run.ended_at}")
    if run.job_name:
        lines.append(f"Pipeline: {run.job_name}")

    return "\n".join(lines)


@click.command(name="get", cls=DgClickCommand, unlaunched=True)
@click.argument("run_id", type=str)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@dg_api_options(deployment_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def get_run_command(
    ctx: click.Context,
    run_id: str,
    output_json: bool,
    organization: str,
    deployment: str,
    api_token: str,
    view_graphql: bool,
) -> None:
    """Get run metadata by ID."""
    from dagster_dg_cli.api_layer.api.run import DgApiRunApi

    config = DagsterPlusCliConfig.create_for_deployment(
        deployment=deployment,
        organization=organization,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    api = DgApiRunApi(client)

    try:
        run = api.get_run(run_id)

        if output_json:
            output = run.model_dump_json(indent=2)
        else:
            output = format_run_table(run)

        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)


@click.group(
    name="run",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "get": get_run_command,
    },
)
def run_group():
    """Manage runs in Dagster Plus."""
