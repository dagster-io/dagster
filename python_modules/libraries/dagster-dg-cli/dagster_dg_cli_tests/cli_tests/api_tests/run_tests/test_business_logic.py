"""Test run business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

import json

from dagster_dg_cli.api_layer.schemas.run import DgApiRun, DgApiRunStatus


def format_run_table(run) -> str:
    """Format run as human-readable table."""
    lines = [
        f"Run ID: {run.id}",
        f"Status: {run.status}",
        f"Created: {run.created_at}",
    ]

    if run.started_at:
        lines.append(f"Started: {run.started_at}")
    if run.ended_at:
        lines.append(f"Ended: {run.ended_at}")
    if run.job_name:
        lines.append(f"Pipeline: {run.job_name}")

    return "\n".join(lines)


class TestFormatRun:
    """Test the run formatting functions."""

    def _create_sample_run(self):
        """Create sample Run for testing."""
        return DgApiRun(
            id="123e4567-e89b-12d3-a456-426614174000",
            status=DgApiRunStatus.SUCCESS,
            created_at=1705311000.0,  # 2024-01-15T10:30:00Z
            started_at=1705311060.0,  # 2024-01-15T10:31:00Z
            ended_at=1705311900.0,  # 2024-01-15T10:45:00Z
            job_name="my_pipeline",
        )

    def _create_minimal_run(self):
        """Create minimal Run with only required fields."""
        return DgApiRun(
            id="minimal-run-123",
            status=DgApiRunStatus.QUEUED,
            created_at=1705311000.0,  # 2024-01-15T10:30:00Z
        )

    def _create_failed_run(self):
        """Create failed Run for testing."""
        return DgApiRun(
            id="failed-run-456",
            status=DgApiRunStatus.FAILURE,
            created_at=1705311000.0,  # 2024-01-15T10:30:00Z
            started_at=1705311060.0,  # 2024-01-15T10:31:00Z
            ended_at=1705311180.0,  # 2024-01-15T10:33:00Z
            job_name="failing_pipeline",
        )

    def _create_canceled_run(self):
        """Create canceled Run for testing."""
        return DgApiRun(
            id="canceled-run-789",
            status=DgApiRunStatus.CANCELED,
            created_at=1705311000.0,  # 2024-01-15T10:30:00Z
            started_at=1705311060.0,  # 2024-01-15T10:31:00Z
            ended_at=1705311120.0,  # 2024-01-15T10:32:00Z
            job_name="canceled_pipeline",
        )

    def test_format_run_text_output(self, snapshot):
        """Test formatting run as text."""
        run = self._create_sample_run()
        result = format_run_table(run)

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_run_json_output(self, snapshot):
        """Test formatting run as JSON."""
        run = self._create_sample_run()
        result = run.model_dump_json(indent=2)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_minimal_run_text_output(self, snapshot):
        """Test formatting minimal run as text."""
        run = self._create_minimal_run()
        result = format_run_table(run)

        snapshot.assert_match(result)

    def test_format_minimal_run_json_output(self, snapshot):
        """Test formatting minimal run as JSON."""
        run = self._create_minimal_run()
        result = run.model_dump_json(indent=2)

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_failed_run_text_output(self, snapshot):
        """Test formatting failed run as text."""
        run = self._create_failed_run()
        result = format_run_table(run)

        snapshot.assert_match(result)

    def test_format_failed_run_json_output(self, snapshot):
        """Test formatting failed run as JSON."""
        run = self._create_failed_run()
        result = run.model_dump_json(indent=2)

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_canceled_run_text_output(self, snapshot):
        """Test formatting canceled run as text."""
        run = self._create_canceled_run()
        result = format_run_table(run)

        snapshot.assert_match(result)

    def test_format_canceled_run_json_output(self, snapshot):
        """Test formatting canceled run as JSON."""
        run = self._create_canceled_run()
        result = run.model_dump_json(indent=2)

        parsed = json.loads(result)
        snapshot.assert_match(parsed)


class TestRunDataProcessing:
    """Test processing of run data structures.

    This class would test any pure functions in the GraphQL adapter
    that process the raw GraphQL responses into our domain models.
    """

    def test_run_creation_with_all_fields(self, snapshot):
        """Test creating run with all possible fields."""
        run = DgApiRun(
            id="complete-run-xyz",
            status=DgApiRunStatus.STARTING,
            created_at=1705311000.0,  # 2024-01-15T10:30:00Z
            started_at=1705311060.0,  # 2024-01-15T10:31:00Z
            ended_at=None,  # Still running
            job_name="comprehensive_pipeline",
        )

        # Test JSON serialization works correctly
        result = run.model_dump_json(indent=2)
        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_run_status_enum_values(self):
        """Test that all expected RunStatus enum values are available."""
        expected_statuses = [
            "QUEUED",
            "STARTING",
            "STARTED",
            "SUCCESS",
            "FAILURE",
            "CANCELING",
            "CANCELED",
        ]

        actual_statuses = [status.value for status in DgApiRunStatus]
        assert set(actual_statuses) == set(expected_statuses)

    def test_run_with_missing_optional_fields(self):
        """Test run creation with None values for optional fields."""
        run = DgApiRun(
            id="sparse-run-123",
            status=DgApiRunStatus.QUEUED,
            created_at=1705311000.0,  # 2024-01-15T10:30:00Z
            started_at=None,
            ended_at=None,
            job_name=None,
        )

        assert run.id == "sparse-run-123"
        assert run.status == DgApiRunStatus.QUEUED
        assert run.created_at == 1705311000.0
        assert run.started_at is None
        assert run.ended_at is None
        assert run.job_name is None
