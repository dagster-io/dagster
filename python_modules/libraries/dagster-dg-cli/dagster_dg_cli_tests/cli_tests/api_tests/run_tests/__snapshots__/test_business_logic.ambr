# serializer version: 1
# name: TestFormatRun.test_format_canceled_run_json_output[0]
  dict({
    'created_at': 1705311000.0,
    'ended_at': 1705311120.0,
    'id': 'canceled-run-789',
    'job_name': 'canceled_pipeline',
    'started_at': 1705311060.0,
    'status': 'CANCELED',
  })
# ---
# name: TestFormatRun.test_format_canceled_run_text_output[0]
  '''
  Run ID: canceled-run-789
  Status: DgApiRunStatus.CANCELED
  Created: 1705311000.0
  Started: 1705311060.0
  Ended: 1705311120.0
  Pipeline: canceled_pipeline
  '''
# ---
# name: TestFormatRun.test_format_failed_run_json_output[0]
  dict({
    'created_at': 1705311000.0,
    'ended_at': 1705311180.0,
    'id': 'failed-run-456',
    'job_name': 'failing_pipeline',
    'started_at': 1705311060.0,
    'status': 'FAILURE',
  })
# ---
# name: TestFormatRun.test_format_failed_run_text_output[0]
  '''
  Run ID: failed-run-456
  Status: DgApiRunStatus.FAILURE
  Created: 1705311000.0
  Started: 1705311060.0
  Ended: 1705311180.0
  Pipeline: failing_pipeline
  '''
# ---
# name: TestFormatRun.test_format_minimal_run_json_output[0]
  dict({
    'created_at': 1705311000.0,
    'ended_at': None,
    'id': 'minimal-run-123',
    'job_name': None,
    'started_at': None,
    'status': 'QUEUED',
  })
# ---
# name: TestFormatRun.test_format_minimal_run_text_output[0]
  '''
  Run ID: minimal-run-123
  Status: DgApiRunStatus.QUEUED
  Created: 1705311000.0
  '''
# ---
# name: TestFormatRun.test_format_run_json_output[0]
  dict({
    'created_at': 1705311000.0,
    'ended_at': 1705311900.0,
    'id': '123e4567-e89b-12d3-a456-426614174000',
    'job_name': 'my_pipeline',
    'started_at': 1705311060.0,
    'status': 'SUCCESS',
  })
# ---
# name: TestFormatRun.test_format_run_text_output[0]
  '''
  Run ID: 123e4567-e89b-12d3-a456-426614174000
  Status: DgApiRunStatus.SUCCESS
  Created: 1705311000.0
  Started: 1705311060.0
  Ended: 1705311900.0
  Pipeline: my_pipeline
  '''
# ---
# name: TestRunDataProcessing.test_run_creation_with_all_fields[0]
  dict({
    'created_at': 1705311000.0,
    'ended_at': None,
    'id': 'complete-run-xyz',
    'job_name': 'comprehensive_pipeline',
    'started_at': 1705311060.0,
    'status': 'STARTING',
  })
# ---
